webapp/src/utils/playbook_run_updates.ts

```diff
@@ -0,0 +1,379 @@
+// Copyright (c) 2020-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {PlaybookRun} from 'src/types/playbook_run';
+import {Checklist, ChecklistItem} from 'src/types/playbook';
+import {TimelineEvent} from 'src/types/rhs';
+import {
+    ChecklistItemUpdatePayload,
+    ChecklistUpdate,
+    ChecklistUpdatePayload,
+    PlaybookRunUpdate,
+} from 'src/types/websocket_events';
+
+// Helper function to apply incremental updates idempotently
+export function applyIncrementalUpdate(currentRun: PlaybookRun, update: PlaybookRunUpdate): PlaybookRun {
+    // Create a new run object with the timestamp update
+    let updatedRun = {
+        ...currentRun,
+        ...(update.playbook_run_updated_at && {update_at: update.playbook_run_updated_at}),
+    };
+
+    // Apply checklist deletions first
+    if (update.checklist_deletes && update.checklist_deletes.length > 0) {
+        const deleteSet = new Set(update.checklist_deletes);
+        updatedRun = {
+            ...updatedRun,
+            checklists: updatedRun.checklists.filter((checklist) => checklist.id && !deleteSet.has(checklist.id)),
+        };
+    }
+
+    // Apply the changed fields to get a new run
+    updatedRun = applyChangedFields(updatedRun, update.changed_fields);
+
+    return updatedRun;
+}
+
+// Helper function to apply changed fields to a playbook run
+function applyChangedFields(run: PlaybookRun, changedFields: PlaybookRunUpdate['changed_fields']): PlaybookRun {
+    // Start with a shallow copy of the run
+    let updatedRun = {...run};
+
+    // Apply simple field changes first
+    for (const [field, value] of Object.entries(changedFields)) {
+        // Skip special fields that need custom handling
+        if (field === 'checklists' || field === 'timeline_events') {
+            continue;
+        }
+
+        // Apply the change to the run using type-safe property assignment
+        if (field in updatedRun) {
+            updatedRun = {
+                ...updatedRun,
+                [field]: value,
+            };
+        }
+    }
+
+    // Handle timeline events specially by merging them with existing events
+    if (changedFields.timeline_events) {
+        updatedRun = applyTimelineUpdates(updatedRun, changedFields.timeline_events);
+    }
+
+    // Apply checklist updates if provided by the server
+    if (changedFields.checklists) {
+        updatedRun = applyChecklistUpdates(updatedRun, changedFields.checklists);
+    }
```

Comment:
Jumping over to array apis isn't necessary IMO; I think a destructuring and a rest+spread approach would be simpler.

```suggestion
    const {timeline_events, checklists, ...basicFields} = changedFields;

    // Start with a shallow copy of the run with basic fields updated
    let updatedRun = {...run, ...basicFields};

    // Handle timeline events specially by merging them with existing events
    if (timeline_events) {
        updatedRun = applyTimelineUpdates(updatedRun, timeline_events);
    }

    // Apply checklist updates if provided by the server
    if (checklists) {
        updatedRun = applyChecklistUpdates(updatedRun, checklists);
    }
```